evalml/pipelines/components/transformers/preprocessing/stl_decomposer.py

```diff
@@ -158,43 +170,62 @@ def fit(self, X: pd.DataFrame, y: pd.Series = None) -> STLDecomposer:
             ValueError: If y is None.
             ValueError: If target data doesn't have DatetimeIndex AND no Datetime features in features data
         """
-        self.original_index = y.index if y is not None else None
-        X, y = self._check_target(X, y)
-        self._map_dt_to_integer(self.original_index, y.index)
-
         # Warn for poor decomposition use with higher seasonal smoothers
         if self.seasonal_smoother > 14:
             self.logger.warning(
                 f"STLDecomposer may perform poorly on data with a high seasonal smoother ({self.seasonal_smoother}).",
             )
 
+        if isinstance(y, pd.Series):
+            y = y.to_frame()
+
+        self.original_index = y.index if y is not None else None
+
+        X, y = self._check_target(X, y)
+
+        self._map_dt_to_integer(self.original_index, y.index)
+
         # Save the frequency of the fitted series for checking against transform data.
         self.frequency = y.index.freqstr or pd.infer_freq(y.index)
-
-        # Determine the period of the seasonal component
-        if self.period is None:
-            self.set_period(X, y)
-
-        stl = STL(y, seasonal=self.seasonal_smoother, period=self.period)
-        res = stl.fit()
-        self.seasonal = res.seasonal
-        self.period = stl.period
-        dist = len(y) % self.period
-        self.seasonality = (
-            self.seasonal[-(dist + self.period) : -dist]
-            if dist > 0
-            else self.seasonal[-self.period :]
-        )
-        self.trend = res.trend
-        self.residual = res.resid
+        # Iterate through each id group
+        self.seasonals = {}
+        self.periods = {}
+        self.seasonalities = {}
+        self.trends = {}
+        self.residuals = {}
+        for id in y.columns:
+            series_y = y[id]
+
+            # Determine the period of the seasonal component
+            if id not in self.periods or self.period is None:
```

Comment:
Yeah, I should have taken `self.period` out. I think I kept it to keep the functionality of `set_period` the same, but is the function used in the `PolynomialDecomposer` at all because I see that it is tested on both decomposers in `test_decomposer_set_period`. Also would the `periods` parameter be same for the single series case or would it be fine for period to be an integer for single series and then take in a dictionary for multiseries?

Reply:
In order to allow users to still pass in a period for both multivariate and univariate cases, I was thinking of keeping both `self.period` and `self.periods` as parameters, but that probably isn't the most efficient implementation so I'm open to other suggestions 😅